src/Mod/Sketcher/Gui/Command.cpp

```diff
@@ -978,16 +977,304 @@ void CmdSketcherViewSection::activated(int iMsg)
+    eType = 0;
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");
+    sPixmap = hGrp->GetBool("GridAuto") ? "Sketcher_GridAuto" : "Sketcher_GridAuto_Deactivated";
+}
+
+void CmdSketcherGridAuto::activated(int iMsg)
+{
+    Q_UNUSED(iMsg);
+    bool newState = toggleGridAuto(getActiveGuiDocument());
+    getAction()->setIcon(Gui::BitmapFactory().iconFromTheme(newState ? "Sketcher_GridAuto" : "Sketcher_GridAuto_Deactivated"));
+}
+
+bool CmdSketcherGridAuto::isActive()
+{
+    auto* sketchView = dynamic_cast<SketcherGui::ViewProviderSketch*>(getActiveGuiDocument()->getInEdit());
+    if (sketchView)
+        getAction()->setIcon(Gui::BitmapFactory().iconFromTheme(sketchView->GridAuto.getValue() ? "Sketcher_GridAuto" : "Sketcher_GridAuto_Deactivated"));
+
+    return isEditModeToolActive(getActiveGuiDocument());
+}
+
+/* Grid comp */
+DEF_STD_CMD_ACL(CmdSketcherCompGrid)
+
+CmdSketcherCompGrid::CmdSketcherCompGrid()
+    : Command("Sketcher_CompGrid")
+{
+    sAppModule = "Sketcher";
+    sGroup = "Sketcher";
+    sMenuText = QT_TR_NOOP("Activate Grid");
+    sToolTipText = QT_TR_NOOP("Activate grid and further settings");
+    sWhatsThis = "Sketcher_CompGrid";
+    sStatusTip = sToolTipText;
+    eType = 0;
+}
+
+void CmdSketcherCompGrid::activated(int iMsg)
+{
+    Gui::ActionGroup* pcAction = qobject_cast<Gui::ActionGroup*>(getAction());
+    if (!pcAction)
+        return;
+    QList<QAction*> a = pcAction->actions();
+
+    if (iMsg == 0) {
+        bool newState = toggleGrid(getActiveGuiDocument());
+        a[0]->setChecked(newState);
+        //activate/deactivate others
+        a[1]->setEnabled(newState);
+        a[2]->setEnabled(newState);
+    }
+    else if (iMsg == 1) {
+        toggleGridSnap(getActiveGuiDocument());
+        pcAction->setProperty("defaultAction", QVariant(0));
+    }
+    else if (iMsg == 2){
+        bool newState = toggleGridAuto(getActiveGuiDocument());
+        a[3]->setEnabled(!newState);
+        pcAction->setProperty("defaultAction", QVariant(0));
+    }
+    else if (iMsg == 3) {
+        //nothing but don't return either.
+    }
+    else
+        return;
+
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");
+    pcAction->setIcon(Gui::BitmapFactory().iconFromTheme(hGrp->GetBool("ShowGrid") ? "Sketcher_GridToggle" : "Sketcher_GridToggle_Deactivated"));
+}
+
+Gui::Action* CmdSketcherCompGrid::createAction()
+{
+    Gui::ActionGroup* pcAction = new Gui::ActionGroup(this, Gui::getMainWindow());
+    pcAction->setDropDownMenu(true);
+    pcAction->setExclusive(false);
+    applyCommandData(this->className(), pcAction);
+
+    QAction* a0 = pcAction->addAction(QString());
+    QAction* a1 = pcAction->addAction(QString());
+    QAction* a2 = pcAction->addAction(QString());
+
+    auto* sizeLabel = new QLabel(QString::fromLatin1("Size"));

```

Comment:
Creating the widget inside this function causes a limitation of this command. If the user decides to use this command for a custom toolbar then the 4th item in the drop-down menu appears only in one instance of the command.

Reply:
I see. But where should it be created then?

Reply:
According to the Qt doc you must create a sub-class of QWidgetAction and implement the createWidget() function
```

class GridSpaceAction : public QWidgetAction
{
public:
    GridSpaceAction(QObject* parent) : QWidgetAction(parent) {
    }

protected:
    QWidget* createWidget(QWidget* parent) override
    {
        auto* sizeLabel = new QLabel(QString::fromLatin1("Size"));
        auto* gridSizeBox = new Gui::QuantitySpinBox();
        gridSizeBox->setProperty("unit", QVariant(QString::fromLatin1("mm")));
        gridSizeBox->setObjectName(QString::fromLatin1("gridSize"));
        gridSizeBox->setToolTip(QApplication::translate("Sketcher_GridSize", "Distance between two subsequent grid lines"));
        //Is General/GridSize correct? From taskSketcherGeneral it seems yes.
        ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General/GridSize");
        gridSizeBox->setValue(hGrp->GetFloat("GridSize", 10.0));
        gridSizeBox->setSingleStep(1.0);
        gridSizeBox->setMaximum(99999999.0);
        gridSizeBox->setMinimum(0.001);
        gridSizeBox->setDecimals(3);

        QObject::connect(gridSizeBox, qOverload<double>(&Gui::QuantitySpinBox::valueChanged),
            [=](double val) {
                Gui::Document* doc = Gui::Application::Instance->activeDocument();
                assert(doc);
                auto* sketchView = dynamic_cast<SketcherGui::ViewProviderSketch*>(doc->getInEdit());
                assert(sketchView);

                sketchView->GridSize.setValue(val);
        });

        QWidget* gridSizeW = new QWidget(parent);
        QHBoxLayout* layout = new QHBoxLayout(gridSizeW);
        layout->addWidget(sizeLabel);
        layout->addWidget(gridSizeBox);
        gridSizeW->installEventFilter(this);
        return gridSizeW;
    }
    bool eventFilter(QObject *obj, QEvent *event) override
    {
        if (event->type() == QEvent::Show) {
            Gui::Document* doc = Gui::Application::Instance->activeDocument();
            if (doc) {
                auto* sketchView = dynamic_cast<SketcherGui::ViewProviderSketch*>(doc->getInEdit());
                if (sketchView) {
                    if (sketchView->GridSize.getValue() > 0) {
                        double gridSize = sketchView->GridSize.getValue();

                        QList<QWidget *> widgets = createdWidgets();
                        for (auto it : widgets) {
                            if (auto* gridSizeBox = it->findChild<Gui::QuantitySpinBox*>(QString::fromLatin1("gridSize"))) {
                                QSignalBlocker sigblk(gridSizeBox);
                                gridSizeBox->setValue(gridSize);
                            }
                        }
                    }
                }
            }
        }

        return QWidgetAction::eventFilter(obj, event);
    }
};

```

Reply:
I tested and it works properly with your code. I have no clue what you mean by the last comment though.